[jit] Support interface python assignment as an attribute

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #27236 [WIP] module swapping as attribute in recursive API
* #26770 [jit] Add interface/object serialization as module attribute
* **#26734 [jit] Support interface python assignment as an attribute**

Summary:

This PR added the python assignment for interface as an attribute in the
module, it enables any object that implicitly inheriting the specific
interface to be able to be assigned to the interface type in python.

Serialization support for interface/class assignment will be done in the
follow up PR

Differential Revision: [D17742708](https://our.internmc.facebook.com/intern/diff/D17742708)

wanchaol committed:
[jit] Support interface python assignment as an attribute

Summary:

This PR added the python assignment for interface as an attribute in the
module, it enables any object that implicitly inheriting the specific
interface to be able to be assigned to the interface type in python.

Serialization support for interface/class assignment will be done in the
follow up PR

wanchaol committed:
Update on "[jit] Support interface python assignment as an attribute"


Summary:

This PR added the python assignment for interface as an attribute in the
module, it enables any object that implicitly inheriting the specific
interface to be able to be assigned to the interface type in python.

Serialization support for interface/class assignment will be done in the
follow up PR

wanchaol committed:
Update on "[jit] Support interface python assignment as an attribute"


Summary:

This PR added the python assignment for interface as an attribute in the
module, it enables any object that implicitly inheriting the specific
interface to be able to be assigned to the interface type in python.

Serialization support for interface/class assignment will be done in the
follow up PR

zdevito commented on test/test_jit.py:
```diff
@@ -19572,8 +19572,37 @@ def wrong4(x):
+            self.assertEqual(scripted_pyassign_mod(input), 2 * input)
+
+            class TestPyAssignError(nn.Module):
+                proxy_mod: OneTwoThree
```
python 3 feature, needs to be in test_jit_py3

wanchaol replied:
seems I need to copy all the relevant classes into test_jit_py3, I will probably just use `__annotations__` in py2 to get this work in this file.

zdevito commented on test/test_jit.py:
```diff
@@ -19572,8 +19572,37 @@ def wrong4(x):
-    # TODO test: interface-interface class-interface inheritance errors,
-    # NamedTuple inheritance errors
+        # Test interface/class python assignment
+        with torch.jit._disable_emit_hooks():
```
_disable_emit_hooks?

zdevito replied:
Is there a bug in current serialization? I would imagine this would already work.

wanchaol replied:
This should be solved by https://github.com/pytorch/pytorch/pull/26770, it's not really a bug but a missing feature, we could be able to serialize the interface type but the run-time object that backed the interface does not get properly serialized, so it missed the `Foo` class when serializing.

zdevito commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -529,8 +529,38 @@ void initJitScriptBindings(PyObject* module) {
             auto unshaped = unshapedType(type);
-            self.register_attribute(
-                name, unshaped, toIValue(std::move(value), type));
+            if (type->cast<InterfaceType>()) {
```
This logic is not in the right place.

The `toIValue(an_class, the_interface_type)` should handle the conversion of the class into an IValue. There are other places where toIValue is called and they should work with interfaces. This only makes it work for attribute assignment.

zdevito commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -529,8 +529,38 @@ void initJitScriptBindings(PyObject* module) {
+                    "a TorchScript compatible type, did you forget to",
+                    "turn it into a user defined TorchScript class?"));
+              }
+              if (!classType->isSubtypeOf(type)) {
```
This check should not be necessary. It should be handled for all types in register_attribute. Furthermore (if it isn't  already), it should be using isSubtypeExt so that it provides more details about why something doesn't match an interface.

zdevito replied:
Actually this check should not be necessary because we should just error in toIValue if the interface type is not compatible with the classes type.

wanchaol replied:
Yeah I will put the `isSubtypeOf` check logic and error out in the `toIValue` instead of here.

zdevito reviewed (commented):
This looks pretty good. The conversion logic should move to toIValue. We should make sure register_attirbute always check and reports subtyping errors.

wanchaol committed:
address comments on "[jit] Support interface python assignment as an attribute"


Summary:

This PR added the python assignment for interface as an attribute in the
module, it enables any object that implicitly inheriting the specific
interface to be able to be assigned to the interface type in python.

Serialization support for interface/class assignment will be done in the
follow up PR

wanchaol committed:
rebase and address comments on "[jit] Support interface python assignment as an attribute"


Summary:

This PR added the python assignment for interface as an attribute in the
module, it enables any object that implicitly inheriting the specific
interface to be able to be assigned to the interface type in python.

Serialization support for interface/class assignment will be done in the
follow up PR

zdevito reviewed (approved):
Looks good, thanks!

wanchaol committed:
Update on "[jit] Support interface python assignment as an attribute"


Summary:

This PR added the python assignment for interface as an attribute in the
module, it enables any object that implicitly inheriting the specific
interface to be able to be assigned to the interface type in python.

Serialization support for interface/class assignment will be done in the
follow up PR

wanchaol committed:
rebase on "[jit] Support interface python assignment as an attribute"


Summary:

This PR added the python assignment for interface as an attribute in the
module, it enables any object that implicitly inheriting the specific
interface to be able to be assigned to the interface type in python.

Serialization support for interface/class assignment will be done in the
follow up PR

wanchaol committed:
Update on "[jit] Support interface python assignment as an attribute"


Summary:

This PR added the python assignment for interface as an attribute in the
module, it enables any object that implicitly inheriting the specific
interface to be able to be assigned to the interface type in python.

Serialization support for interface/class assignment will be done in the
follow up PR

Differential Revision: [D17742708](https://our.internmc.facebook.com/intern/diff/D17742708)
